File: app/src/aether/tools/list_directory.py
"""List Directory tool — show files in a directory."""

from __future__ import annotations

import os

from aether.tools.base import AetherTool, ToolParam, ToolResult

MAX_ENTRIES = 200


class ListDirectoryTool(AetherTool):
    name = "list_directory"
    description = "List files and subdirectories in the given directory path. Shows names, types (file/dir), and sizes."
    status_text = "Listing files..."
    parameters = [
        ToolParam(name="path", type="string", description="Directory path to list", required=False, default="."),
    ]

    def __init__(self, working_dir: str = "."):
        self.working_dir = os.path.abspath(working_dir)

    def _resolve(self, path: str) -> str:
        resolved = os.path.abspath(os.path.join(self.working_dir, path))
        if not resolved.startswith(self.working_dir):
            raise PermissionError(f"Path escapes working directory: {path}")
        return resolved

    async def execute(self, path: str = ".") -> ToolResult:
        resolved = self._resolve(path)

        if not os.path.exists(resolved):
            return ToolResult.fail(f"Directory not found: {path}")

        if not os.path.isdir(resolved):
            return ToolResult.fail(f"Not a directory: {path}")

        try:
            entries = sorted(os.listdir(resolved))
            if len(entries) > MAX_ENTRIES:
                entries = entries[:MAX_ENTRIES]
                truncated = True
            else:
                truncated = False

            lines = []
            for name in entries:
                full = os.path.join(resolved, name)
                if os.path.isdir(full):
                    lines.append(f"  {name}/")
                else:
                    size = os.path.getsize(full)
                    lines.append(f"  {name}  ({size:,} bytes)")

            output = "\n".join(lines) if lines else "(empty directory)"
            if truncated:
                output += f"\n... truncated at {MAX_ENTRIES} entries"

            return ToolResult.success(output, path=path, count=len(entries))
        except PermissionError:
            return ToolResult.fail(f"Permission denied: {path}")
        except Exception as e:
            return ToolResult.fail(f"Error listing directory: {e}")

File: agent/internal/tools/builtin/list_directory.go
package builtin

import (
	"context"
	"os"
	"sort"
	"strings"

	"github.com/suryaumapathy2812/core-ai/agent/internal/tools"
)

type ListDirectoryTool struct{}

func (t *ListDirectoryTool) Definition() tools.Definition {
	return tools.Definition{
		Name:        "list_directory",
		Description: "List files and directories under a path within the working directory.",
		StatusText:  "Listing directory...",
		Parameters:  []tools.Param{{Name: "path", Type: "string", Description: "Relative path to list", Required: false, Default: "."}},
	}
}

func (t *ListDirectoryTool) Execute(ctx context.Context, call tools.Call) tools.Result {
	_ = ctx
	path, _ := call.Args["path"].(string)
	resolved, err := resolveSafePath(call.Ctx.WorkingDir, path)
	if err != nil {
		return tools.Fail("Invalid path: "+err.Error(), nil)
	}
	entries, err := os.ReadDir(resolved)
	if err != nil {
		return tools.Fail("Failed to list directory: "+err.Error(), nil)
	}
	lines := make([]string, 0, len(entries))
	for _, e := range entries {
		name := e.Name()
		if e.IsDir() {
			name += "/"
		}
		lines = append(lines, name)
	}
	sort.Strings(lines)
	return tools.Success(strings.Join(lines, "\n"), map[string]any{"count": len(lines), "path": resolved})
}

var _ tools.Tool = (*ListDirectoryTool)(nil)
